### PR TITLE
Signup: create site preview iframe component

### PR DESCRIPTION
**Pull request by ramonjd**
## Changes proposed in this Pull Request
This PR introduces an asynchronously-loaded component that encapsulates a site preview, with mocked content and current theme CSS for both mobile and phone views.

Apr-10-2019 17-13-44

This works by creating an iframe containing a WordPress page HTML shell into which we pump:

1. a theme CSS URI
2. a Google Font CSS URI
3. title, tagline and body content

## TODO
- [x] Make it work in IE11 :trollface: 
Screen Shot 2019-04-09 at 5 38 06 pm

## Testing instructions

Fire up the branch, test the preview. In particular, play around with http://calypso.localhost:3000/start/onboarding-dev/site-style-with-preview

Preview in different widths, throttle the network, try to break it.

For bonus points, apply the yet-unfinished vertical annotation localization patch `D26776-code` and switch to a `RTL` language such as Hebrew or Arabic.

The iframe HTML tag should update to something like: `<html lang="he" dir="rtl">` and look right-to-left:

Screen Shot 2019-04-12 at 12 36 56 pm

### CODE REVIEW

**Comment by fditrapani** [comment]
Thanks for putting this together ❤️ 

I noticed we lost some detailing around the devices:
image

Notice the height of the mobile device and the missing line under the phone label. I don't mind the scrolling and height change on the desktop browser but I think we should maintain phone-like proportions to make it look right. 

image

The type on sophisticated looks really huge too. Can we fake it a little and reduce the size so that people can get a better sense of what the theme looks like? Otherwise it's all text.

Also cc'ing @shaunandrews to take a look.

**Comment by shaunandrews** [comment]
> I don't mind the scrolling and height change on the desktop browser but I think we should maintain phone-like proportions to make it look right.

Agreed on the phone — the proportions are important to making it obvious that this is a phone preview. However, I disagree on the desktop preview height and scrolling. I feel the max-height and overflow scrolling just creates another scrollable area, and makes it a little difficult to scroll.

--

image

- I'm seeing the border as a single pixel in Safari. It's should be `3px` everywhere.
- We should use the `--color-accent-light` variable for the border color.
- It doesn't seem that I'm able to `tab` around to select options. Can we add keyboard navigation?
- Just like other steps, let's switch from the "Continue" label to the arrow icon for the button.
- The containing element should get an `elevation ( 2dp )` to get the right box-shadow.
- Similar, the button should get an `elevation ( 1dp )` to get a box-shadow.
- Related: I seems the previews have lost their box-shadow. Can we bring that back?

File: client/components/site-mockup/style.scss [hunk 1]
```diff
@@ -1,382 +0,0 @@
...
-// Mockup document body; Where all the
-// stuff goes.
-.site-mockup__body {
-	height: 100%;
-	overflow: hidden;
-
-	// Mobile body is scrollable and needs to
-	// make room for the chrome.
-	.site-mockup__viewport.is-mobile & {
-		border-top: 1px solid var( --color-neutral-100 );
```
Comment (michaeldcain) [inline client/components/site-mockup/style.scss]: I think this is the missing phone chrome border.

**Comment by ramonjd** [comment]
> The type on sophisticated looks really huge too. 

I noticed this too, so I checked against a site I created with this theme applied. According to the theme CSS, it's the intended size for widths above `768px`.

Screen Shot 2019-04-10 at 10 34 42 am

Rather than fudge the text size on our size, for a true representation of what the user will see after signup, we can give the desktop window a max width of `<= 768px`:

Screen Shot 2019-04-10 at 10 34 31 am

**Comment by ramonjd** [comment]
Thanks @shaunandrews and @fditrapani for the feedback. 

Most of these I think we can take care of; many were due to silly rebase errors.

> I disagree on the desktop preview height and scrolling. I feel the max-height and overflow scrolling just creates another scrollable area, and makes it a little difficult to scroll.

Okay, because we no longer control the styles in the Calypso app, we'll probably have to calculate the height of the internal iframe window and dynamically set the container height accordingly. I'll give it a shot

> I'm seeing the border as a single pixel in Safari. It's should be 3px everywhere...

I see this too on Safari in production. I think these and other changes related to the site style buttons/unrelated to the preview windows are best completed in another PR. @shaunandrews feel free to work on this in the meantime, otherwise I can look at them later.

> I seems the previews have lost their box-shadow. Can we bring that back?

👍

**Comment by ramonjd** [comment]
> Notice the height of the mobile device and the missing line under the phone label. 

✅

**Comment by fditrapani** [comment]
This is looking really good. Thanks for all the updates.

**Comment by michaeldcain** [comment]
The 768px max-width works really well for the fonts, but we lose the overflowing image detail in the media+text block for the Modern theme at that resolution. If it's one or the other, let's go with the cover image text.

Screen Shot 2019-04-11 at 10 48 14 AM

**Comment by ramonjd** [comment]
> The 768px max-width works really well for the fonts, but we lose the overflowing image detail in the media+text block for the Modern theme at that resolution. If it's one or the other, let's go with the cover image text.

No worries. I'll revert the desktop container back to the larger width. 

If unreasonably huge text is undesirable then I'd say that we can and should adjust it in the theme, rather than playing bait and switch in signup.

**Comment by ramonjd** [comment]
I've had to bite the digital bullet and overwrite the height value for `.wp-block-cover` on the Calypso side due to the js height assignment mentioned in https://github.com/Automattic/wp-calypso/pull/32135#issuecomment-482354201

`<SignupSitePreview />` could probably be abstracted more, particularly the CSS, to provide more control should we use it elsewhere. A good task for another day.

**Comment by fditrapani** [comment]
> <SignupSitePreview /> could probably be abstracted more, particularly the CSS, to provide more control should we use it elsewhere. A good task for another day.

I big plus one for another day :D 

This is really coming along nicely! Thanks for all your work here @ramonjd. Can't wait to get this out and also swap the default theme to Modern — it looks brighter and fresher than professional.

**Comment by ramonjd** [comment]
@cain @adamziel 

Would you mind giving a final test of this? This PR adds site mockup to the `onboarding-dev` flow only.

I've spent the morning on manual testing. All looks well in various browsers. 

It might be my slow internet connection today, but IE 11/Edge on BrowserStack intermittently applied the wrong media queries in the mobile iframe view

Screen Shot 2019-04-17 at 12 02 41 pm

Here, IE is applying a 60px right and left margin to `entry-content`. The margin is meant only for screen widths wider than 768.

As I said, it only happens now and then, and I can't nail down the cause. I suspect some sort of race condition. 

If it's a huge problem I can try to preload all the assets in another PR, though we already do a window.onload on the frame... 🤔

File: client/components/signup-site-preview/utils.js [hunk 2]
```diff
@@ -0,0 +1,139 @@
+/** @format */
+
+/**
+ * Internal dependencies
+ */
+
+// Detect IE 11 and below
+export function isIE() {
+	const ua = window.navigator.userAgent;
+	return -1 !== ua.indexOf( 'MSIE' ) || -1 !== ua.indexOf( 'Trident/' );
```
Comment (adamziel) [inline client/components/signup-site-preview/utils.js]: Looks like we have some code that does this in `client/lib/user-agent/index.js` - it's just a note and not a blocker though

**Comment by fditrapani** [comment]
I took this for a quick spin and it looks good. 

I don't know if this was introduced in this PR but I noticed a weird flash when I move from step to step. It's like the fields are unloading and then loading in abruptly. Can anyone else confirm this behaviour on their end? If it's an issue, it can probably be addressed in another PR though.

File: client/signup/site-mockup/index.jsx [hunk 3]
```diff
@@ -137,53 +127,86 @@ class SiteMockups extends Component {
...
+				'site-topic-with-preview' === ownProps.stepName ||
+				'site-information-title-with-preview' === ownProps.stepName,
+			themeSlug: style.theme,
+			fontUrl: style.fontUrl,
+		};
+	},
+	dispatch => ( {
+		handleClick: ( verticalSlug, siteStyle, size ) =>
+			dispatch(
+				recordTracksEvent( 'calypso_signup_site_preview_mockup_clicked', {
```
Comment (adamziel) [inline client/signup/site-mockup/index.jsx]: I don't think this event is recorded - is the click handler getting attached properly?

**Comment by adamziel** [comment]
Actually I found one problem, see my comment above. I'm okay with fixing it in another PR

File: client/components/signup-site-preview/utils.js (hunk 2 above)
Comment (ramonjd) [reply client/components/signup-site-preview/utils.js]: I did include that library first then realized I didn't need the entire lib, just a one-line test for IE.

File: client/components/signup-site-preview/iframe.jsx [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+			this.props.scrolling
+		);
+
+		if ( isIE() ) {
+			this.iframe.current.contentWindow.document.open();
+			this.iframe.current.contentWindow.document.write( iframeSrc );
+			this.iframe.current.contentWindow.document.close();
+		} else {
+			// For memory management: https://developer.mozilla.org/en-US/docs/Web/API/URL/revokeObjectURL
+			URL.revokeObjectURL( this.iframe.current.src );
```
[the remark this answers is not in the document]
Comment (ramonjd) [reply client/components/signup-site-preview/iframe.jsx]: > (URL.createObjectURL) is encapsulated in getIframeSource but it's referred to also outside of it.

Good observation!!!

I think I'd like to abstract `URL.revokeObjectURL` out of the component, so I might push it over to `utils.js` via another util function to keep the implementation in one place.

**Comment by ramonjd** [comment]
@fditrapani @adamziel Thanks again for testing this!!

> I don't know if this was introduced in this PR but I noticed a weird flash when I move from step to step. It's like the fields are unloading and then loading in abruptly. Can anyone else confirm this behaviour on their end? If it's an issue, it can probably be addressed in another PR though.

I noticed this yesterday too, and speculate that it might be due to the recent PR #32282, which updated [client/signup/config/step-components.js](https://github.com/Automattic/wp-calypso/pull/32282/files#diff-902f02ea4f5154c176be5f9abfba4189) so that each step loads asynchronously.

#32282 reduces the initial signup bundles by a significant amount, so it'd be preferable to reach a balance. 

We might be able to mitigate the effect by some visual trickery, that is, displaying a heading and a placeholder block.

Or perhaps preload the next step somehow. 

cc @sgomes

File: client/signup/site-mockup/index.jsx (hunk 3 above)
Comment (ramonjd) [reply client/signup/site-mockup/index.jsx]: It's triggered by a click on the iframe body. 

Works for me:

Apr-18-2019 07-40-05

Any way I can reproduce what you're experiencing?

I can fix directly after merge if it persists.

**Comment by fditrapani** [comment]
> I noticed this yesterday too, and speculate that it might be due to the recent PR #32282, which updated client/signup/config/step-components.js so that each step loads asynchronously.

Ok good to know — Thanks @ramonjd . We can tackle that in another PR/issue. I created a Trello card and mentioned you in it  for any other ideas.

**Comment by ramonjd** [comment]
This is probably ready to merge, but the `test-e2e-canary-ie` are failing across the board.

**Comment by sgomes** [comment]
> @fditrapani @adamziel Thanks again for testing this!!
> 
> > I don't know if this was introduced in this PR but I noticed a weird flash when I move from step to step. It's like the fields are unloading and then loading in abruptly. Can anyone else confirm this behaviour on their end? If it's an issue, it can probably be addressed in another PR though.
> 
> I noticed this yesterday too, and speculate that it might be due to the recent PR #32282, which updated [client/signup/config/step-components.js](https://github.com/Automattic/wp-calypso/pull/32282/files#diff-902f02ea4f5154c176be5f9abfba4189) so that each step loads asynchronously.
> 
> #32282 reduces the initial signup bundles by a significant amount, so it'd be preferable to reach a balance.
> 
> We might be able to mitigate the effect by some visual trickery, that is, displaying a heading and a placeholder block.
> 
> Or perhaps preload the next step somehow.
> 
> cc @sgomes

Hello!

Yes, the flash sounds like something that may have been introduced by my PR, since the required data to render the step is no longer immediately available, requiring a network fetch first (this is how it avoids loading everything upfront). I apologise if that's causing problems 🙁 

I'm happy to help in any efforts to minimise these visual glitches. Is there a good way for me to test this without creating new accounts all the time?

File: client/signup/site-mockup/index.jsx (hunk 3 above)
Comment (adamziel) [reply client/signup/site-mockup/index.jsx]: All good, I think it was just my ublock :-)

**Comment by fditrapani** [comment]
Thanks @sgomes — no problem. I created an issue in Trello for us to work on the solution outside of this PR. I'll send you a link over Slack so you can stay up to date. 

> Is there a good way for me to test this without creating new accounts all the time?

Once you create an account, you can continue to create sites under one account.

**Comment by ramonjd** [comment]
Thanks @sgomes ! 

`D20305-code` also makes it easier to test the signup journey, removing the need to sign in or create a new user all the time in incognito mode :)

File: client/signup/site-mockup/index.jsx (hunk 3 above)
Comment (ramonjd) [reply client/signup/site-mockup/index.jsx]: Block all that WordPress spam! Hehe

**Comment by fditrapani** [comment]
Yay! thanks for all your work on this @ramonjd.

**Pull request merged**